cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -0,0 +1,249 @@
+
+    ProgramStatusNotificationSubscriberThread(String topic) {
+      super(topic);
+      this.topic = topic;
+    }
+
+    @Override
+    public String loadMessageId(DatasetContext context) {
+      return getAppMetadataStore(context).retrieveSubscriberState(topic);
+    }
+
+    @Override
+    public void updateMessageId(DatasetContext context, String lastFetchedMessageId) {
+      getAppMetadataStore(context).persistSubscriberState(topic, lastFetchedMessageId);
+    }
+
+    @Override
+    public void processNotification(final DatasetContext context, Notification notification)
+      throws IOException, DatasetManagementException {
+      Map<String, String> properties = notification.getProperties();
+      // Required parameters
+      String programRunIdString = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+      String programStatusString = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
+
+      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus
+      if (programRunIdString == null || programStatusString == null) {
+        return;
+      }
+
+      ProgramRunStatus programRunStatus;
+      try {
+        programRunStatus = ProgramRunStatus.valueOf(programStatusString);
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Invalid program run status {} passed in notification for program {}",
+                 programStatusString, programRunIdString);
+        return;
+      }
+
+      ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      ProgramId programId = programRunId.getParent();
+
+      ApplicationMeta meta = getAppMetadataStore(context).getApplication(programRunId.getNamespace(),
+                                                                         programRunId.getApplication(),
+                                                                         programRunId.getVersion());
+      // Check if the application exists
+      if (meta == null) {
+        LOG.warn(new NotFoundException(programId.getParent()).getMessage());
+        return;
+      }
+      // Check if the program exists
+      if (getProgramSpecFromApp(meta.getSpec(), programRunId) == null) {
+        LOG.warn(new NotFoundException(programId).getMessage());
+        return;
+      }
+
+      String runId = programRunId.getRun();
+      String twillRunId = notification.getProperties().get(ProgramOptionConstants.TWILL_RUN_ID);
+      long endTimeSecs = getTimeSeconds(notification.getProperties(), ProgramOptionConstants.END_TIME);
+      switch(programRunStatus) {
+        case STARTING:
+          long startTimeSecs = getTimeSeconds(notification.getProperties(), ProgramOptionConstants.START_TIME);
+          String userArgumentsString = properties.get(ProgramOptionConstants.USER_OVERRIDES);
+          String systemArgumentsString = properties.get(ProgramOptionConstants.SYSTEM_OVERRIDES);
+          if (userArgumentsString == null || systemArgumentsString == null) {
```

Comment:
It's the same reason we check for an invalid ProgramRunId or ProgramRunStatus in the initial notification. Just like programRunId or programRunStatus in the notification, it can't really be null.